### ISSUE TITLE
Static or dynamic loading refactor

### ISSUE THREAD

**Issue by Beanow** [sourcecred/sourcecred#1506]
## UniRef Context

In order to support [Unified Reference Detection](https://discourse.sourcecred.io/t/unified-reference-detection/301) we will have a dependency on our plugin's mirrored data in the createGraph stage.

Specifically, the reference detector should only return a "hit" for URLs that exists in our mirror. And to make it unified, every plugin will depend on every other plugin. E.g. GitHub's createGraph will depend on Discourse, GitHub and Initiatives data.

## Splitting load logic

Currently in `/src/api/load.js` we're treating each plugin's load as a single step which combines the mirroring and createGraph. Based on the new requirement the idea was to split this into two stages.

1. Populating the data layer of plugins (mirroring / syncing with APIs).
2. Graph creation.

In between those two steps, we can create our reference detector with data dependencies and pass that on to the graph creation functions. There's one exception to this, node contractions should always be last, so we'll consider this another stage.

```js
export type PopulationInput = {|
  +githubToken: string | null,
  +project: Project,
  +cacheDirectory: string,
  +taskReporter: TaskReporter,
|};

export type PopulationResult = {|
  +referenceDetector?: ReferenceDetector,
  +discourseData?: ReadRepository,
  +createGraph: GraphCreationFunction,
|};

export type GraphCreationInput = {|
  +referenceDetector: ReferenceDetector,
  +discourseData: ?ReadRepository,
|};

export type PopulateFunction = (PopulationInput) => ?Promise<PopulationResult>;

// Note: we're not providing the PopulationInput again. We're expecting necessary data
// to be provided through for example closures.
export type GraphCreationFunction = (GraphCreationInput) => Promise<Graph>;

async function stagedLoading(
  populateFns: $ReadOnlyArray<PopulateFunction>,
  input: PopulationInput
): Promise<Graph> {
  // First stage: parallel population

  // Reducing stage 1 results to stage 2 input

  // Second stage: graph creation.

  // Final stage: identity node contractions.
}
```

## Scaling flaw in this abstraction

So this sort of works in a generic way, but the Initiatives plugin wouldn't fit these stages.
The way the Initiatives reference detector would work is:
1. Use the Discourse data to create our set of `Initiative` objects and build a
`trackerToInitiative: Map<NodeAddressT, NodeAddressT>` based on this.
1. Use the Discourse reference detector to not need any knowledge about URL format and instead handle the `NodeAddressT` coming from the Discourse reference detector.
1. "Upgrade" any hits from the Discourse reference detector to the Initiative node address using our map. Returning null if there's no mapping.

So clearly we have a dependency here on Discourse's data and reference detector. But we have this dependency in Initiative's stage 1, to create a reference detector. **A same-stage dependency.**

In this particular instance, we could mitigate this problem by making yet another stage.
1. Populating the data layer of plugins (mirroring / syncing with APIs).
2. Creating reference detectors.
2. Graph creation.
4. Node contraction.

But if we're being honest here, this is a bandaid. Because our Initiatives reference detector makes use of the Initiatives data layer (`InitiativeRepository`) here. So if we ever wanted to depend on this from another plugin, we're back to having a **same-stage dependency** between the Discourse and Initiatives data layers.
https://github.com/sourcecred/sourcecred/blob/91de7cfbc7c06ef281cd74cdce8d0ed6f9521b94/src/plugins/initiatives/initiative.js#L32-L40

Taken to it's logical extreme, a stages pattern is not able to fully capture dependency structures.

## Secondary issue, static type dependencies

Notice how these supposed generic types have very plugin specific fields, like:
`githubToken: string | null` and `discourseData?: ReadRepository`

This will cause our core types to have a dependency on every plugin under the sun that might want to expose something to other plugins. That is, if we want to use Flow's static analysis to validate these types and functions. This issue also exists at the configuration layer, especially the `Project` type has a similar situation.

To create a dynamic plugin system, we'll run into this problem more often that we may not have access to the Flow types and runtime type casting would make more sense to generalize plugin integration points like these.

## Alternative A: static types + hard-coded loading

The simplest approach would be to punt on the plugin system and a dynamic handling of dependencies by not attempting to generalize it now. Instead of using abstractions to create stages, fully hard-code the loading dependencies for the plugins we have right now.

This allows us to eliminate the same-stage dependency issue by simply hand crafting the load order. And we can fully use the Flow types to our advantage by creating new types and function signatures to reflect exactly the inputs we need and return types we're depending on.

## Alternative B: runtime types + generalized loading

The generalized loading structure could be more accurately represented as a DAG. Something I suggested as part of https://github.com/sourcecred/sourcecred/issues/1425

However to do so safely, we would need to look into runtime type checking as well.
Perhaps we could use something like [`tcomb`](https://www.npmjs.com/package/tcomb) to accept an `any` Flow type, validate it with `tcomb` and cast it to a corresponding Flow type? :thinking:

---

Currently I'm leaning towards A, and have built a prototype with that fairly quickly that does everything I need. It's fairly simple and in https://github.com/sourcecred/sourcecred/issues/1425 we already thought we should punt on this area.

Thoughts? @decentralion @wchargin

**Comment by decentralion** [comment]
Thanks for the explanation @Beanow! I understand the same-stage loading problem now.

I agree that we should take alternative A. Having the plugin system "really working" (i.e. external developers can add plugins without needing to change core) is not a goal for the beta. IMO it's a task for 2021 or later. Adding more technical debt for something that's more than a year out, in order to unblock the next year's worth of goals, makes a lot of sense.

**Comment by Beanow** [comment]
Alternative A has been implemented through #1586

Mainly visible in the [`PluginLoaders`](https://github.com/sourcecred/sourcecred/blob/master/src/backend/pluginLoaders.js) and the plugin Loader interfaces it utilizes, which remain unique per plugin.
